hs_core/models.py

```diff
@@ -1750,70 +1766,494 @@ class Meta:
+
+        There are two main usages to this constructor:
+
+        * uploading a file from a form or REST call:
+
+                ResourceFile.create(r, File(...something...), folder=d)
+
+        * copying a file internally from iRODS:
+
+                ResourceFile.create(r, file_name, folder=d, source=s, move=True)
+          or
+                ResourceFile.create(r, file_name, folder=d, source=s, move=False)
+
+        In this case, source is a full iRODS pathname of the place from which to copy or move
+        the file. The default is to copy the file and leave a copy in place.
+
+        A third form is less common and presumes that the file already exists in iRODS:
+
+        * pointing to an existing file:
+
+                ResourceFile.create(r, file_name, folder=d)
+
+        """
+        # bind to appropriate resource
+        kwargs = {}
+        if __debug__:
+            assert isinstance(resource, BaseResource)
+        kwargs['content_object'] = resource
+
+        kwargs['file_folder'] = folder
+
+        # if file is an open file, use native copy by setting appropriate variables
+        if isinstance(file, File):
+            if resource.resource_federation_path:
+                kwargs['resource_file'] = None
+                kwargs['fed_resource_file'] = file
+            else:
+                kwargs['resource_file'] = file
+                kwargs['fed_resource_file'] = None
+
+        else:  # if file is not an open file, then it's a basename (string)
+            if file is None and source is not None:
+                root, file = os.path.split(source)  # take from source path
+            target = get_resource_path(resource, file, folder=folder)
+            if source is not None:
+                istorage = resource.get_irods_storage()
+                if not move:
+                    istorage.copyFiles(source, target)
+                else:
+                    istorage.moveFile(source, target)
+            # we've copied or moved if necessary; now set the paths
+            if resource.resource_federation_path:
+                kwargs['resource_file'] = None
+                kwargs['fed_resource_file'] = target
+            else:
+                kwargs['resource_file'] = target
+                kwargs['fed_resource_file'] = None
+
+        # Actually create the file record
+        # when file is a File, the file is copied to storage in this step
+        # otherwise, the copy must precede this step.
+        return ResourceFile.objects.create(**kwargs)
+
+    # TODO: automagically handle orphaned logical files
+    def delete(self):
+        """
+        Delete a resource file record and the file contents
+
+        model.delete does not cascade to delete files themselves,
+        and these must be explicitly deleted.
+
+        """
+        if self.exists:
+            if self.fed_resource_file:
+                self.fed_resource_file.delete()
```

Comment:
@alvacouch I see what you are doing now, and it looks like using fed_resource_file can cover the third scenario I mentioned above. However, we will need to make sure fed_resource_file to contain the full absolute iRODS path including irods zone in order for federated irods storage to delete the file properly.